doc/content/design/host-ntp-time.md

```diff
@@ -0,0 +1,224 @@
+
+On the host, dhclient executes `/etc/dhcp/dhclient.d/chrony.sh` to update the
+ntp servers when network event happens.
+- `chrony.sh` writes ntp servers to `/run/chrony-dhcp/$interface.sources`
+- Chonryd include the dir `/run/chrony-dhcp` by `sourcedir /run/chrony-dhcp` in
+the conf file
+- `chrony.sh` runs `chronyc reload sources` to reload NTP sources
+
+Then NTP sources can be updated automatically in the DHCP mode
+
+How to switch DHCP mode?
+
+Dhclient stores dhcp lease in `/var/lib/xcp/dhclient-$interface.leases`, see
+module Dhclient in `/ocaml/networkd/lib/network_utils.ml`.
+
+When switch the NTP mode to DHCP, XAPI
+- check ntp server item in the lease and fills it in chrony-dhcp files
+- Add the exec permission of `chrony.sh`
+- Remove all the NTP source items(custom or default) in chronyd conf
+- Restart chronyd
+
+When swith ntp mode from dhcp to others, XAPI
+- Remove the chrony-dhcp files
+- Remove the exec permission of chrony.sh
+- Add NTP source items(custom or default) in chronyd conf
+- Restart chronyd
+
+### Custom mode
+
+In this mode, NTP uses `host.ntp_custom_servers` as sources.
+
+When switch the NTP mode to custom, XAPI
+- Remove NTP source items in chronyd conf
+- Add `host.ntp_custom_servers` as NTP source items in chronyd conf
+- Resetart chronyd
+
+When `host.ntp_custom_servers` changes and `host.mode` is custom, set chronyd
+conf with new custom servers and restart chronyd.
+
+### Default mode
+
+In this mode, ntp uses `default-ntp-servers` in XAPI config file. Generally the
+default-ntp-servers will be defined by the product.
+
+### Others
+
+`host.enable/disable_ntp` starts or stops chronyd service.
+
+`host.get_ntp_servers_status` calls `chronyc -c sources` to get ntp servers status.
+Output parse:
+```
+Source mode: '^' = server, '=' = peer, '#' = local clock.
+Source state: '*' = current synced, '+' = combined, '-' = not combined,
+'?' = unreachable, 'x' = time may be in error, '~' = time too variable
+```
+
+## Timezone configuration on the host
+
+New filed: `host.timezone`, string
+
+New APIs: `host.set_timezone`, `host.get_timezone`, `host.list_timezones`
+
+The timezone is in IANA timezone database format. Timezone on the host can be
+get by `realpath /etc/localtime` which is linked to timezone file under
+`/usr/share/zoneinfo/`. It can set by link`/etc/localtime` to
+`/usr/share/zoneinfo/<timezone>`. All the valid timezones are actually the files
+under `/usr/share/zoneinfo/`.
+
+Comparing to using a fixed UTC offset, the benefit is:
+- User-friendly: familiar region names and same with most system facilities.
+- Handles daylight saving time (DST) automatically
+
+They are equivalent to the `timedatectl` commands
+```
+timedatectl set-timezone
+timedatectl status | grep "Time zone"
+timedatectl list-timezones
+```
+
+## Time on the host
+
+New API: `host.get_ntp_synchronized`, `host.set_server_localtime`
+
+`host.get_ntp_synchronized` shows if the system time is synchronized with NTP
```

Comment:
This can be a implemented by a call to `ntp_adjtime`, and checking for `TIME_ERROR`, see https://man7.org/linux/man-pages/man2/adjtimex.2.html
(Note: do not check for `TIME_OK`, because there are a lot of leap-second specific return codes during normal operation, which are not errors).